platform/logging/LoggerFactory.ts

```diff
@@ -0,0 +1,109 @@
+
+export interface LoggerFactory {
+  get(...namespace: string[]): Logger;
+}
+
+export class MutableLoggerFactory implements LoggerFactory {
+  private config?: LoggingConfig;
+  private readonly appenders: Map<string, BaseAppender> = new Map([
+    [BUFFERED_APPENDER_KEY, new BufferAppender(
+      new BaseAppenderConfig({ kind: 'buffer', pattern: '' })
+    )]
+  ]);
+  private readonly loggers: Map<string, LoggerAdapter> = new Map();
+
+  get(...namespace: string[]): Logger {
+    let context = namespace.join('.');
+    if (this.loggers.has(context)) {
+      return this.loggers.get(context)!;
+    }
+
+    let loggerConfig;
+    if (this.config) {
+      loggerConfig = this.getLoggerConfigByContext(this.config, context);
+    } else {
+      // If we don't have config yet, use `buffered` appender that will store all logged messages in the memory
+      // until the config is ready.
+      loggerConfig = { level: LogLevel.All, appenders: [BUFFERED_APPENDER_KEY] };
+    }
+
+    this.loggers.set(context,
+      new LoggerAdapter(
+        new BaseLogger(
+          namespace,
+          loggerConfig.level,
+          loggerConfig.appenders.map((appenderKey) => this.appenders.get(appenderKey)!)
+        )
+      )
+    );
+
+    return this.loggers.get(context)!;
+  }
+
+  updateConfig(config: LoggingConfig) {
+    // TODO: Should we support `config.isEqual()` to avoid updating the config if it's still the same?
+    this.config = config;
+
+    // Before we dispose appenders, let's check whether we have any `buffered` appender that may need to be
+    // flushed via newly configured appenders.
+    const bufferedLogRecords = this.appenders.has(BUFFERED_APPENDER_KEY) ?
+      [...(<BufferAppender>this.appenders.get(BUFFERED_APPENDER_KEY)!).buffer] :
+      [];
+
+    this.updateAppendersFromConfig(config);
+    this.updateLoggersFromConfig(config);
+
+    // Re-log all buffered log records with newly configured appenders.
+    for (const logRecord of bufferedLogRecords) {
+      this.get(logRecord.context.join('.')).log(logRecord);
+    }
+  }
+
+  close(): void {
+    for (const appender of this.appenders.values()) {
+      appender.close();
+    }
+
+    this.appenders.clear();
+    this.loggers.clear();
+  }
+
+  private updateAppendersFromConfig(config: LoggingConfig) {
+    for (const appender of this.appenders.values()) {
+      appender.close();
+    }
+
+    this.appenders.clear();
+
+    // Initialize new log appenders.
+    for (const [appenderKey, appenderConfig] of config.appenders.entries()) {
+      this.appenders.set(appenderKey, appenderConfig.createAppender());
+    }
+  }
+
+  private updateLoggersFromConfig(config: LoggingConfig) {
+    for (const [loggerKey, loggerAdapter] of this.loggers.entries()) {
+      const loggerConfig = this.getLoggerConfigByContext(config, loggerKey);
+      loggerAdapter.logger = new BaseLogger(
+        // TODO: Do we need to store namespace as an array at all?
```

Comment:
I initially thought there was value in it, but I'm not sure anymore. E.g. for my comment below on `getLoggerConfigByContext` it might even be easier to use strings instead. Personally I feel the array is "cleaner" (by some definition of "clean" 🙈), but if the apis feel better with strings instead (which it looks like right now), I'm +1 on moving to strings